src/python/pants/backend/python/target_types.py

```diff
@@ -648,7 +653,24 @@ class PythonRequirementsFileSources(Sources):
 class PythonRequirementsFile(Target):
     alias = "_python_requirements_file"
     core_fields = (*COMMON_TARGET_FIELDS, PythonRequirementsFileSources)
-    help = "A private, helper target type for requirements.txt files."
+    help = "A private helper target type for requirements.txt files."
+
+
+# -----------------------------------------------------------------------------------------------
+# `_python_constraints` target
+# -----------------------------------------------------------------------------------------------
+
+
+class PythonRequirementConstraintsField(_RequirementSequenceField):
+    alias = "constraints"
+    required = True
+    help = "A list of pip-style requirement strings, e.g. `my_dist==4.2.1`."
+
+
+class PythonRequirementConstraints(Target):
+    alias = "_python_constraints"
+    core_fields = (*COMMON_TARGET_FIELDS, PythonRequirementConstraintsField)
+    help = "A private helper target for inlined requirements constraints, used by macros."
```

Comment:
The expectation with #11165 is that folks will be creating these manually I think? Also, we'll soon support other types of lockfiles than `constraints.txt`.

Is it worth skipping to publicly naming this a `python_lockfile` target (with a `constraints=` field for the sources)?

Reply:
> with a constraints= field for the sources

We do not want a `Sources` field here because the main idea of this PR is that you can create constraints purely through strings in a BUILD file - you don't need any file to exist on disk.  We already have `--requirement-constraints` to read a file on disk, it would be repetitive to have the target do it.

However, I had considered removing the `--requirement-constraints` option to always use the target no matter what. But, we decided to hold off on that given the possible churn for multiple lockfiles.

> Also, we'll soon support other types of lockfiles than constraints.txt.
> Is it worth skipping to publicly naming this a python_lockfile target

I don't think so. It's not clear yet what these other lockfiles will look like, including if we support multiple formats or only one. I want to avoid premature generalization.

But, this does confirm for me that we probably want to keep this target private.

Reply:
> But, this does confirm for me that we probably want to keep this target private.

If the target is private but the macro is public, you still have "something" that is public. I think that my assumption with making it public is that the macro might not be necessary.